### PR TITLE
fix(fsevents): correctly handle unknown events

### PR DESCRIPTION
**Pull request by rgrinberg**
We should apply our action mask before converting it into an OCaml
variant. Otherwise, we'll get events that we don't understand.

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

ps-id: 204d7599-9ffb-4e95-8cff-871f39b4dc68

### CODE REVIEW

**Comment by rgrinberg** [comment]
@samoht could you give this PR a try?

**Comment by samoht** [comment]
Should it be on top of the other one?

**Comment by samoht** [comment]
I still have a `Error: exception Failure("fsevents: unexpected event action")` with just this PR.

**Comment by rgrinberg** [comment]
> Should it be on top of the other one? (if so could you make a branch with all the patches - it's easier to pin/test for me)

There's no need. That PR just improved the error message.

> I still have a Error: exception Failure("fsevents: unexpected event action") with just this PR.

What about now?

File: src/fsevents/fsevents_stubs.c [hunk 1]
```diff
@@ -261,20 +261,22 @@ static const FSEventStreamEventFlags action_mask =
     kFSEventStreamEventFlagItemCreated | kFSEventStreamEventFlagItemRemoved |
     kFSEventStreamEventFlagItemRenamed | kFSEventStreamEventFlagItemModified;
 
+static int count_set_bits(uint32_t bits) {
+  // XXX slow
```
Comment (emillon) [inline src/fsevents/fsevents_stubs.c:265]: I suppose we don't need to be very fast, but `__builtin_popcount(bits)` should do the tricks for the compilers we support on mac.

Comment (rgrinberg) [reply src/fsevents/fsevents_stubs.c:265]: Good call. We can try that out in a separate PR.

**Comment by rgrinberg** [comment]
@emillon could you review the PR? This PR fixes one part of the bug, the other part will be fixed in a subsequent PR.

File: src/fsevents/fsevents_stubs.c [hunk 2]
```diff
@@ -261,20 +261,22 @@ static const FSEventStreamEventFlags action_mask =
...
-  // XXX slow
-  int count = 0;
-  while (flags) {
-    count += (flags & 1);
-    flags >>= 1;
-  }
-
-  flags = Int32_val(v_flags);
-  if (count >= 2 || flags & kFSEventStreamEventFlagItemRenamed) {
+  if (flags & kFSEventStreamEventFlagItemRenamed || count_set_bits(flags) >= 2) {
```
Comment (emillon) [inline src/fsevents/fsevents_stubs.c:279]:
```suggestion
  if ((flags & kFSEventStreamEventFlagItemRenamed) || (count_set_bits(flags) >= 2)) {
```

I always have a double take with operator precedence

File: src/fsevents/fsevents_stubs.c (hunk 1 above)
Comment (emillon) [reply src/fsevents/fsevents_stubs.c:265]: did so in #6224

**Comment by rgrinberg** [comment]
@emillon one more round of review please.

@samoht could you test again?

File: src/fsevents/fsevents_stubs.c [hunk 3]
```diff
@@ -270,24 +270,16 @@ CAMLprim value dune_fsevents_action(value v_flags) {
...
-  flags = Int32_val(v_flags);
-  if (count >= 2 || flags & kFSEventStreamEventFlagItemRenamed) {
-    // we don't bother tracking renamed acurately for now. macos makes it
-    // tricky by not telling is which path is created and which one is deleted.
-    // it is possible to reverse engineer this from the chain of inodes in the
-    // events, but it's also error prone as inodes can be reused. so for now, we
-    // avoid this is and treat renamed as unknown
-    v_action = Val_int(0);
-  } else if (flags & kFSEventStreamEventFlagItemCreated) {
+  if (kFSEventStreamEventFlagItemRenamed) {
```
Comment (emillon) [inline src/fsevents/fsevents_stubs.c:273]:
```suggestion
  if (flags & kFSEventStreamEventFlagItemRenamed) {
```

(`kFS` ... is a constant)

File: src/fsevents/fsevents_stubs.c [hunk 4]
```diff
@@ -270,24 +270,16 @@ CAMLprim value dune_fsevents_action(value v_flags) {
...
     v_action = Val_int(1);
-  } else if (flags & kFSEventStreamEventFlagItemRemoved) {
+  } else if (flags & kFSEventStreamEventFlagItemCreated) {
     v_action = Val_int(2);
-  } else if (flags & kFSEventStreamEventFlagItemModified) {
+  } else if (flags & kFSEventStreamEventFlagItemRemoved) {
     v_action = Val_int(3);
+  } else if (flags & kFSEventStreamEventFlagItemModified) {
+    v_action = Val_int(4);
```
Comment (emillon) [inline src/fsevents/fsevents_stubs.c:280]: looks like the `Val_int`s do not correspond to the flag being tested.

**Review by emillon — CHANGES_REQUESTED** [review]

**Comment by samoht** [comment]
No more crash!

File: src/fsevents/fsevents_stubs.c (hunk 4 above)
Comment (rgrinberg) [reply src/fsevents/fsevents_stubs.c]: Fixed now and I re-ordered the constants to make it easier to add more and check if the order is correct.

**Review by emillon — APPROVED** [review]

**Pull request merged**